File: webapp/libfgg/transport/websocket.ts
import { IConn } from "./conn"

import { encode, decode } from './protocol'

export default class Conn implements IConn {
  conn: WebSocket
  constructor(conn: WebSocket) {
    this.conn = conn
    this.conn.binaryType = "arraybuffer"
  }
  send(head: ArrayBuffer, body: ArrayBuffer): Promise<void> {
    this.conn.send(encode(head, body))
    return new Promise<void>((fn) => fn())
  }
  setOnRecv(fn: (_: ArrayBuffer, __: ArrayBuffer) => void): void {
    this.conn.onmessage = (ev: MessageEvent<ArrayBuffer>) => {
      const [head, body] = decode(ev.data)
      fn(head, body)
    }
  }
}

File: libfgg/transport/websocket/websocket.go
package websocket

import (
	"context"
	"sync"
	"time"

	"filegogo/libfgg/transport/protocol"

	"github.com/gorilla/websocket"

	log "github.com/sirupsen/logrus"
)

type Conn struct {
	conn      *websocket.Conn
	mutex     sync.Mutex
	onMessage func([]byte, []byte)
}

func New(conn *websocket.Conn) *Conn {
	return &Conn{
		conn:      conn,
		onMessage: func([]byte, []byte) {},
	}
}

func (c *Conn) SetOnRecv(fn func(head, body []byte)) {
	c.onMessage = fn
}

func (c *Conn) Run(ctx context.Context) {
	go func() {
		select {
		case <-ctx.Done():
			c.conn.Close()
		}
	}()

	for {
		typ, data, err := c.conn.ReadMessage()
		log.Tracef("WebSocket RECV %d: %s\n", typ, data)
		if err != nil {
			log.Error(err)
		}

		if data == nil {
			time.Sleep(100 * time.Millisecond)
			continue
		}

		c.onMessage(protocol.Decode(data))
	}
}

func (c *Conn) Send(head, body []byte) error {
	data := protocol.Encode(head, body)
	c.mutex.Lock()
	defer c.mutex.Unlock()
	return c.conn.WriteMessage(websocket.BinaryMessage, data)
}
